File: backend/src/services/metrics_service.py
"""Service for aggregating and analyzing product metrics data.

This service provides helper methods for:
- Calculating metric changes and trends
- Comparing multiple products
- Generating summary statistics
- Time-range analysis
"""

from datetime import datetime, timedelta
from decimal import Decimal
from uuid import UUID

from loguru import logger
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from products.models import (
    Product,
    ProductSnapshot,
)
from schemas.metrics import (
    BSRTrendData,
    MetricComparisonResponse,
    MetricDataPoint,
    MetricsSummary,
    PriceTrendData,
    ProductMetricTrend,
    ReviewTrendData,
)


class MetricsAggregationService:
    """Service for aggregating and analyzing product metrics."""

    @staticmethod
    async def get_metrics_summary(db: AsyncSession, product_id: UUID) -> MetricsSummary | None:
        """Get summary statistics for a product with change percentages.

        Args:
            db: AsyncSession for database queries
            product_id: Product ID

        Returns:
            MetricsSummary with current values and 7d/30d changes, or None if no data
        """
        # Get latest metric
        stmt_latest = (
            select(ProductSnapshot)
            .where(
                ProductSnapshot.product_id == product_id,
            )
            .order_by(ProductSnapshot.scraped_at.desc())
            .limit(1)
        )
        result_latest = await db.execute(stmt_latest)
        latest = result_latest.scalar_one_or_none()

        # If no snapshots exist, return None (can't calculate metrics without data)
        if latest is None:
            raise ValueError(f"No snapshots found for product {product_id}")

        # Get metrics from 7 days ago
        date_7d_ago = datetime.utcnow() - timedelta(days=7)
        stmt_7d = (
            select(ProductSnapshot)
            .where(
                ProductSnapshot.product_id == product_id,
                ProductSnapshot.scraped_at <= date_7d_ago,
            )
            .order_by(ProductSnapshot.scraped_at.desc())
            .limit(1)
        )
        result_7d = await db.execute(stmt_7d)
        metric_7d = result_7d.scalar_one_or_none()

        # Get metrics from 30 days ago
        date_30d_ago = datetime.utcnow() - timedelta(days=30)
        stmt_30d = (
            select(ProductSnapshot)
            .where(
                ProductSnapshot.product_id == product_id,
                ProductSnapshot.scraped_at <= date_30d_ago,
            )
            .order_by(ProductSnapshot.scraped_at.desc())
            .limit(1)
        )
        result_30d = await db.execute(stmt_30d)
        metric_30d = result_30d.scalar_one_or_none()

        # Calculate changes (with None checks)
        # Extract values with None checks first
        # Using type: ignore because mypy doesn't properly narrow Optional types in if blocks
        price_7d: Decimal | float | None = None
        if metric_7d is not None:
            price_7d = metric_7d.price  # type: ignore[union-attr]

        price_30d: Decimal | float | None = None
        if metric_30d is not None:
            price_30d = metric_30d.price  # type: ignore[union-attr]

        bsr_7d: int | None = None
        if metric_7d is not None:
            bsr_7d = metric_7d.bsr_main_category  # type: ignore[union-attr]

        bsr_30d: int | None = None
        if metric_30d is not None:
            bsr_30d = metric_30d.bsr_main_category  # type: ignore[union-attr]

        rating_7d: Decimal | float | None = None
        if metric_7d is not None:
            rating_7d = metric_7d.rating  # type: ignore[union-attr]

        rating_30d: Decimal | float | None = None
        if metric_30d is not None:
            rating_30d = metric_30d.rating  # type: ignore[union-attr]

        review_count_7d: int | None = None
        if metric_7d is not None:
            review_count_7d = metric_7d.review_count  # type: ignore[union-attr]

        review_count_30d: int | None = None
        if metric_30d is not None:
            review_count_30d = metric_30d.review_count  # type: ignore[union-attr]

        price_change_7d = MetricsAggregationService._calculate_change(price_7d, latest.price)
        price_change_30d = MetricsAggregationService._calculate_change(price_30d, latest.price)
        bsr_change_7d = MetricsAggregationService._calculate_change(
            bsr_7d, latest.bsr_main_category
        )
        bsr_change_30d = MetricsAggregationService._calculate_change(
            bsr_30d, latest.bsr_main_category
        )
        rating_change_7d = MetricsAggregationService._calculate_change(rating_7d, latest.rating)
        rating_change_30d = MetricsAggregationService._calculate_change(rating_30d, latest.rating)

        review_growth_7d = (
            latest.review_count - review_count_7d if review_count_7d is not None else None
        )
        review_growth_30d = (
            latest.review_count - review_count_30d if review_count_30d is not None else None
        )

        return MetricsSummary(
            product_id=product_id,
            current_price=latest.price,
            price_change_7d=price_change_7d,
            price_change_30d=price_change_30d,
            current_bsr=latest.bsr_main_category,
            bsr_change_7d=bsr_change_7d,
            bsr_change_30d=bsr_change_30d,
            current_rating=latest.rating,
            rating_change_7d=rating_change_7d,
            rating_change_30d=rating_change_30d,
            review_count=latest.review_count,
            review_growth_7d=review_growth_7d,
            review_growth_30d=review_growth_30d,
            last_updated=latest.scraped_at,
        )

    @staticmethod
    async def get_product_comparison(
        db: AsyncSession, product_ids: list[UUID], metric_type: str, days: int = 30
    ) -> MetricComparisonResponse:
        """Get comparison data for multiple products.

        Args:
            db: AsyncSession for database queries
            product_ids: List of product IDs (max 10)
            metric_type: Type of metric ('price', 'bsr', 'rating', 'reviews')
            days: Number of days to include (default 30)

        Returns:
            MetricComparisonResponse with trend data for all products
        """
        start_date = datetime.utcnow() - timedelta(days=days)
        end_date = datetime.utcnow()

        products_data = []

        for product_id in product_ids[:10]:  # Limit to 10 products
            stmt_product = select(Product).where(Product.id == product_id)
            result_product = await db.execute(stmt_product)
            product = result_product.scalar_one_or_none()

            if not product:
                continue

            # Get metrics for date range
            stmt_metrics = (
                select(ProductSnapshot)
                .where(
                    ProductSnapshot.product_id == product_id,
                    ProductSnapshot.scraped_at >= start_date,
                    ProductSnapshot.scraped_at <= end_date,
                )
                .order_by(ProductSnapshot.scraped_at)
            )
            result_metrics = await db.execute(stmt_metrics)
            metrics = result_metrics.scalars().all()

            # Convert to appropriate trend data based on metric type
            data_points: list[PriceTrendData] | list[BSRTrendData] | list[ReviewTrendData]
            if metric_type == "price":
                data_points = [
                    PriceTrendData(
                        date=m.scraped_at,
                        price=float(m.price) if m.price is not None else None,
                        buybox_price=float(m.buybox_price) if m.buybox_price is not None else None,
                        original_price=float(m.original_price)
                        if m.original_price is not None
                        else None,
                        category_avg_price=(
                            float(m.category_avg_price)
                            if m.category_avg_price is not None
                            else None
                        ),
                    )
                    for m in metrics
                ]
            elif metric_type == "bsr":
                data_points = [
                    BSRTrendData(
                        date=m.scraped_at,
                        bsr_main=m.bsr_main_category,
                        bsr_small=m.bsr_small_category,
                    )
                    for m in metrics
                ]
            elif metric_type in ["rating", "reviews"]:
                data_points = [
                    ReviewTrendData(
                        date=m.scraped_at,
                        rating=m.rating,
                        review_count=m.review_count,
                        category_avg_rating=m.category_avg_rating,
                        category_avg_reviews=m.category_avg_reviews,
                    )
                    for m in metrics
                ]
            else:
                data_points = []

            products_data.append(
                ProductMetricTrend(
                    product_id=product.id,
                    product_title=product.title,
                    product_asin=product.asin,
                    data_points=data_points,
                )
            )

        # Get category average if available
        first_product_id = str(product_ids[0]) if product_ids else ""
        category_average = await MetricsAggregationService._get_category_average_trend(
            db,
            first_product_id,
            metric_type,
            start_date,
            end_date,
        )

        return MetricComparisonResponse(
            metric_type=metric_type,
            start_date=start_date,
            end_date=end_date,
            products=products_data,
            category_average=category_average,
        )

    @staticmethod
    async def _get_category_average_trend(
        db: AsyncSession,
        category: str,
        metric_field: str,
        start_date: datetime,
        end_date: datetime,
    ) -> list[MetricDataPoint] | None:
        """Get category average trend for a specific metric."""
        # TODO: Implement category average trend calculation
        return None

    @staticmethod
    def _calculate_change(
        old_value: float | Decimal | None, new_value: float | Decimal | None
    ) -> float | None:
        """Calculate percentage change between two values.

        Args:
            old_value: Previous value
            new_value: Current value

        Returns:
            Percentage change, or None if calculation not possible
        """
        if old_value is None or new_value is None:
            return None

        try:
            old = float(old_value)
            new = float(new_value)

            if old == 0:
                return None

            return ((new - old) / old) * 100
        except Exception as e:
            logger.error(f"Error calculating change: {e}")
            return None

    @staticmethod
    @staticmethod
    async def get_category_trend(
        db: AsyncSession, category_name: str, days: int = 30
    ) -> list[dict]:
        """Get category average trends for the last N days.

        Args:
            db: Database session
            category_name: Name of the category
            days: Number of days to fetch (default: 30)

        Returns:
            List of category metrics with timestamp and values
        """
        # TODO: Implement category trend calculation
        return []

File: frontend/src/services/metricsService.ts
/**
 * Product Metrics Service
 *
 * TypeScript service for fetching and managing product metrics data.
 * Provides methods for trend analysis and multi-product comparison.
 */

import apiClient from '@/lib/api';

// ============================================================================
// Interfaces
// ============================================================================

export interface ProductMetric {
  id: string;
  product_id: string;
  recorded_at: string;
  price: number | null;
  original_price: number | null;
  buybox_price: number | null;
  discount_percentage: number | null;
  bsr_main_category: number | null;
  bsr_small_category: number | null;
  rating: number | null;
  review_count: number;
  in_stock: boolean;
  stock_status: string | null;
  seller_name: string | null;
  is_amazon_seller: boolean;
  is_fba: boolean;
  is_deal: boolean;
  is_prime: boolean;
  coupon_available: boolean;
  coupon_text: string | null;
  category_avg_price: number | null;
  category_avg_rating: number | null;
  category_avg_reviews: number | null;
  scrape_successful: boolean;
  scrape_error: string | null;
}

export interface MetricsSummary {
  product_id: string;
  current_price: number | null;
  price_change_7d: number | null;
  price_change_30d: number | null;
  current_bsr: number | null;
  bsr_change_7d: number | null;
  bsr_change_30d: number | null;
  current_rating: number | null;
  rating_change_7d: number | null;
  rating_change_30d: number | null;
  review_count: number;
  review_growth_7d: number | null;
  review_growth_30d: number | null;
  last_updated: string;
}

export interface PriceTrendDataPoint {
  date: string;
  price: number | null;
  buybox_price: number | null;
  original_price: number | null;
  category_avg_price?: number | null;
}

export interface BSRTrendDataPoint {
  date: string;
  bsr_main: number | null;
  bsr_small: number | null;
}

export interface ReviewTrendDataPoint {
  date: string;
  rating: number | null;
  review_count: number;
  category_avg_rating?: number | null;
  category_avg_reviews?: number | null;
}

export interface MetricDataPoint {
  date: string;
  value: number | null;
}

export interface ProductMetricTrend {
  product_id: string;
  product_title: string;
  product_asin: string;
  data_points:
    | PriceTrendDataPoint[]
    | BSRTrendDataPoint[]
    | ReviewTrendDataPoint[];
}

export interface MetricComparisonResponse {
  metric_type: 'price' | 'bsr' | 'rating' | 'reviews';
  start_date: string;
  end_date: string;
  products: ProductMetricTrend[];
  category_average: MetricDataPoint[] | null;
}

export interface MetricComparisonRequest {
  product_ids: string[];
  metric_type: 'price' | 'bsr' | 'rating' | 'reviews';
  days?: number;
}

export interface CategoryTrendRequest {
  category_name: string;
  days?: number;
}

// ============================================================================
// NEW: Dynamic Metrics System Interfaces
// ============================================================================

export interface MetricFieldMetadata {
  display_name: string;
  description: string;
  type: string;
  category: string;
}

export interface DynamicTrendResponse {
  metadata: Record<string, MetricFieldMetadata>;
  data: Array<Record<string, any>>;
  total_points: number;
}

export interface FieldCategory {
  name: string;
  display_name: string;
  description: string;
  type: string;
}

export interface AvailableFieldsResponse {
  categories: Record<string, FieldCategory[]>;
  total_fields: number;
}

// ============================================================================
// Service Methods
// ============================================================================

export const metricsService = {
  /**
   * Get metrics summary for a product with 7-day and 30-day changes
   */
  async getMetricsSummary(productId: string): Promise<MetricsSummary> {
    const response = await apiClient.get(
      `/api/v1/metrics/products/${productId}/summary`
    );
    return response.data;
  },

  /**
   * Get all metrics for a product within a date range
   */
  async getProductMetrics(
    productId: string,
    days: number = 30
  ): Promise<ProductMetric[]> {
    const response = await apiClient.get(
      `/api/v1/metrics/products/${productId}/metrics`,
      {
        params: { days },
      }
    );
    return response.data;
  },

  /**
   * Get the most recent metric snapshot for a product
   */
  async getLatestMetric(productId: string): Promise<ProductMetric> {
    const response = await apiClient.get(
      `/api/v1/metrics/products/${productId}/latest`
    );
    return response.data;
  },

  // ============================================================================
  // NEW: Dynamic Metrics API
  // ============================================================================

  /**
   * Get all available metric fields that can be queried
   *
   * **Returns:** Field registry with categories, types, and descriptions
   *
   * @example
   * const fields = await metricsService.getAvailableFields();
   * // Build UI dropdown from fields.categories
   */
  async getAvailableFields(): Promise<AvailableFieldsResponse> {
    const response = await apiClient.get('/api/v1/metrics/fields/available');
    return response.data;
  },

  /**
   * 🚀 Dynamic metric trends endpoint - Query any combination of fields!
   *
   * **This replaces getPriceTrend(), getBSRTrend(), and getReviewTrend()**
   *
   * @param productId - Product ID to query
   * @param fields - Array of field names to retrieve
   * @param days - Number of days to go back (1-365)
   *
   * @example
   * // Get price data (replaces getPriceTrend)
   * const priceData = await metricsService.getTrends(123, ['price', 'buybox_price', 'original_price'], 30);
   *
   * @example
   * // Get BSR data (replaces getBSRTrend)
   * const bsrData = await metricsService.getTrends(123, ['bsr_main', 'bsr_small'], 30);
   *
   * @example
   * // Get review data (replaces getReviewTrend)
   * const reviewData = await metricsService.getTrends(123, ['rating', 'review_count'], 30);
   *
   * @example
   * // Get multiple metrics in one call
   * const dashboardData = await metricsService.getTrends(
   *   123,
   *   ['price', 'bsr_main', 'rating', 'review_count', 'in_stock'],
   *   90
   * );
   */
  async getTrends(
    productId: string,
    fields: string[],
    days: number = 30
  ): Promise<DynamicTrendResponse> {
    const response = await apiClient.get(
      `/api/v1/metrics/products/${productId}/trends`,
      {
        params: {
          fields: fields.join(','),
          days,
        },
      }
    );
    return response.data;
  },

  /**
   * Get price trend data for a product (backward compatible wrapper)
   *
   * @deprecated Use getTrends() instead for better flexibility
   * @example
   * // Old way:
   * const data = await metricsService.getPriceTrend(123, 30);
   *
   * // New way (recommended):
   * const data = await metricsService.getTrends(123, ['price', 'buybox_price', 'original_price'], 30);
   */
  async getPriceTrend(
    productId: string,
    days: number = 30
  ): Promise<PriceTrendDataPoint[]> {
    const response = await this.getTrends(
      productId,
      ['price', 'buybox_price', 'original_price'],
      days
    );
    return response.data.map((point: any) => ({
      date: point.date,
      price: point.price,
      buybox_price: point.buybox_price,
      original_price: point.original_price,
    }));
  },

  /**
   * Get BSR trend data for a product (backward compatible wrapper)
   *
   * @deprecated Use getTrends() instead for better flexibility
   * @example
   * // Old way:
   * const data = await metricsService.getBSRTrend(123, 30);
   *
   * // New way (recommended):
   * const data = await metricsService.getTrends(123, ['bsr_main', 'bsr_small'], 30);
   */
  async getBSRTrend(
    productId: string,
    days: number = 30
  ): Promise<BSRTrendDataPoint[]> {
    const response = await this.getTrends(
      productId,
      ['bsr_main', 'bsr_small'],
      days
    );
    return response.data.map((point: any) => ({
      date: point.date,
      bsr_main: point.bsr_main,
      bsr_small: point.bsr_small,
    }));
  },

  /**
   * Get review count and rating trend for a product (backward compatible wrapper)
   *
   * @deprecated Use getTrends() instead for better flexibility
   * @example
   * // Old way:
   * const data = await metricsService.getReviewTrend(123, 30);
   *
   * // New way (recommended):
   * const data = await metricsService.getTrends(123, ['rating', 'review_count'], 30);
   */
  async getReviewTrend(
    productId: string,
    days: number = 30
  ): Promise<ReviewTrendDataPoint[]> {
    const response = await this.getTrends(
      productId,
      ['rating', 'review_count'],
      days
    );
    return response.data.map((point: any) => ({
      date: point.date,
      rating: point.rating,
      review_count: point.review_count,
    }));
  },

  /**
   * Compare metrics for multiple products with optional category average overlay
   */
  async compareProducts(
    request: MetricComparisonRequest
  ): Promise<MetricComparisonResponse> {
    const response = await apiClient.post('/api/v1/metrics/compare', request);
    return response.data;
  },

  /**
   * Get category average trend data
   */
  async getCategoryTrend(
    categoryName: string,
    days: number = 30
  ): Promise<MetricDataPoint[]> {
    const response = await apiClient.post('/api/v1/metrics/category/trend', {
      category_name: categoryName,
      days,
    });
    return response.data;
  },
};

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Format percentage change with + or - sign
 */
export function formatPercentageChange(value: number | null): string {
  if (value === null || value === undefined) {
    return 'N/A';
  }
  const sign = value > 0 ? '+' : '';
  return `${sign}${value.toFixed(2)}%`;
}

/**
 * Get color for percentage change (green for positive, red for negative)
 * Note: For BSR, lower is better, so colors are inverted
 */
export function getChangeColor(
  value: number | null,
  invertColors: boolean = false
): string {
  if (value === null || value === undefined) {
    return 'text-gray-500';
  }

  const isPositive = value > 0;
  const isGood = invertColors ? !isPositive : isPositive;

  return isGood ? 'text-green-600' : 'text-red-600';
}

/**
 * Format currency value
 */
export function formatCurrency(
  value: number | null,
  currency: string = 'USD'
): string {
  if (value === null || value === undefined) {
    return 'N/A';
  }

  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency,
  }).format(value);
}

/**
 * Format BSR (Best Seller Rank) with commas
 */
export function formatBSR(value: number | null): string {
  if (value === null || value === undefined) {
    return 'N/A';
  }

  return `#${new Intl.NumberFormat('en-US').format(value)}`;
}

/**
 * Format rating (1-5 stars)
 */
export function formatRating(value: number | null): string {
  if (value === null || value === undefined) {
    return 'N/A';
  }

  return `${value.toFixed(1)} ★`;
}

/**
 * Get metric type display name
 */
export function getMetricTypeName(metricType: string): string {
  const names: Record<string, string> = {
    price: 'Price',
    bsr: 'Best Seller Rank',
    rating: 'Rating',
    reviews: 'Review Count',
  };

  return names[metricType] || metricType;
}

/**
 * Get chart color by index for multi-product comparison
 */
export function getChartColor(index: number): string {
  const colors = [
    '#3b82f6', // blue
    '#10b981', // green
    '#f59e0b', // amber
    '#ef4444', // red
    '#8b5cf6', // purple
    '#ec4899', // pink
    '#14b8a6', // teal
    '#f97316', // orange
    '#6366f1', // indigo
    '#84cc16', // lime
  ];

  return colors[index % colors.length];
}
